File: ios-app/RifatCamPro/StreamServer.swift
import UIKit

class StreamServer {
    private var serverFd: Int32 = -1
    private var running = false
    private var clients: Set<Int32> = []
    private let clientsLock = NSLock()
    private let serverQueue = DispatchQueue(label: "stream.server.queue", qos: .userInitiated)
    private var latestJPEG: Data?
    private let jpegLock = NSLock()

    var deviceName = "RifatCam iPhone"
    var port: UInt16 = NetworkUtils.streamPort

    func updateJPEG(_ data: Data) {
        jpegLock.lock()
        latestJPEG = data
        jpegLock.unlock()
    }

    func start() {
        guard !running else { return }
        running = true

        serverQueue.async { [weak self] in
            self?.runServer()
        }
    }

    func stop() {
        running = false
        clientsLock.lock()
        for fd in clients {
            close(fd)
        }
        clients.removeAll()
        clientsLock.unlock()
        if serverFd >= 0 {
            close(serverFd)
            serverFd = -1
        }
    }

    var clientCount: Int {
        clientsLock.lock()
        let count = clients.count
        clientsLock.unlock()
        return count
    }

    private func runServer() {
        serverFd = Darwin.socket(AF_INET, SOCK_STREAM, 0)
        guard serverFd >= 0 else { print("[Server] socket() failed"); return }

        var opt: Int = 1
        setsockopt(serverFd, SOL_SOCKET, SO_REUSEADDR, &opt, socklen_t(MemoryLayout.size(ofValue: opt)))

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = CFSwapInt16HostToBig(port)
        addr.sin_addr.s_addr = INADDR_ANY
        addr.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)

        let bindResult = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                Darwin.bind(serverFd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }

        guard bindResult == 0 else {
            print("[Server] bind() failed on port \(port)")
            close(serverFd)
            serverFd = -1
            return
        }

        listen(serverFd, 5)
        print("[Server] Listening on port \(port)")

        while running {
            var clientAddr = sockaddr_in()
            var addrLen = socklen_t(MemoryLayout<sockaddr_in>.size)
            let clientFd = withUnsafeMutablePointer(to: &clientAddr) {
                $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                    Darwin.accept(serverFd, $0, &addrLen)
                }
            }

            guard clientFd >= 0 else {
                if running { print("[Server] accept() error") }
                break
            }

            clientsLock.lock()
            clients.insert(clientFd)
            clientsLock.unlock()

            DispatchQueue.global(qos: .default).async { [weak self] in
                self?.handleClient(clientFd)
            }
        }

        close(serverFd)
        serverFd = -1
        print("[Server] Stopped")
    }

    private func handleClient(_ fd: Int32) {
        defer {
            close(fd)
            clientsLock.lock()
            clients.remove(fd)
            clientsLock.unlock()
        }

        var buffer = [UInt8](repeating: 0, count: 4096)
        let bytesRead = read(fd, &buffer, 4096)
        guard bytesRead > 0 else { return }

        let request = String(bytes: buffer[0..<min(bytesRead, 4096)], encoding: .utf8) ?? ""
        let lines = request.components(separatedBy: "\r\n")
        guard let firstLine = lines.first else { return }

        if firstLine.hasPrefix("GET /video") {
            handleVideoStream(fd)
        } else if firstLine.hasPrefix("GET /info") {
            handleInfo(fd)
        } else if firstLine.hasPrefix("POST /control") {
            handleControl(fd, request: request)
        } else if firstLine.hasPrefix("GET / ") || firstLine == "GET / HTTP/1.0" || firstLine == "GET / HTTP/1.1" {
            handleIndex(fd)
        } else {
            sendResponse(fd, statusCode: 404, contentType: "text/plain", body: "Not Found")
        }
    }

    private func handleVideoStream(_ fd: Int32) {
        let headers = "HTTP/1.0 200 OK\r\n" +
            "Content-Type: multipart/x-mixed-replace; boundary=--boundary\r\n" +
            "Cache-Control: no-cache, no-store, must-revalidate\r\n" +
            "Pragma: no-cache\r\n" +
            "Connection: close\r\n" +
            "Access-Control-Allow-Origin: *\r\n\r\n"
        writeString(fd, headers)

        let boundaryPrefix = "--boundary\r\n" +
            "Content-Type: image/jpeg\r\n" +
            "Content-Length: %d\r\n\r\n"

        while running {
            var jpeg: Data?
            jpegLock.lock()
            jpeg = latestJPEG
            jpegLock.unlock()

            if let data = jpeg {
                let header = String(format: boundaryPrefix, data.count)
                writeString(fd, header)
                data.withUnsafeBytes { ptr in
                    if let base = ptr.baseAddress {
                        _ = write(fd, base, data.count)
                    }
                }
                writeString(fd, "\r\n")
            } else {
                Thread.sleep(forTimeInterval: 0.03)
            }
        }
    }

    private func handleInfo(_ fd: Int32) {
        let ip = NetworkUtils.getWiFiAddress() ?? "unknown"
        let json = """
        {"status":"streaming","name":"\(deviceName)","device_id":"ios-1","protocol":"mjpeg","ip":"\(ip)","port":\(port)}
        """
        sendResponse(fd, statusCode: 200, contentType: "application/json", body: json)
    }

    private func handleControl(_ fd: Int32, request: String) {
        let body = request.components(separatedBy: "\r\n\r\n").last ?? "{}"
        let response = "{\"status\":\"ok\",\"received\":\(body.count > 0 ? "true" : "false")}"
        sendResponse(fd, statusCode: 200, contentType: "application/json", body: response)
    }

    private func handleIndex(_ fd: Int32) {
        let ip = NetworkUtils.getWiFiAddress() ?? "localhost"
        let html = """
        <!DOCTYPE html>
        <html>
        <head>
            <title>RifatCam Pro - \(deviceName)</title>
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <style>
                * { margin:0; padding:0; box-sizing:border-box; }
                body { background:#0a0e17; color:#fff; font-family:-apple-system, sans-serif; display:flex; flex-direction:column; align-items:center; justify-content:center; min-height:100vh; }
                h1 { color:#00d4ff; font-size:22px; letter-spacing:3px; margin-bottom:16px; }
                img { width:100%%; max-width:720px; border:1px solid rgba(0,212,255,0.3); border-radius:12px; }
                .info { color:#8899aa; font-size:13px; margin-top:16px; }
                .status { color:#00ff88; font-size:14px; margin-top:8px; }
            </style>
        </head>
        <body>
            <h1>&#9670; RIFATCAM PRO</h1>
            <img src="/video" alt="Live Stream" />
            <div class="info">Device: \(deviceName) | \(ip):\(port)</div>
            <div class="status">&#9679; LIVE</div>
        </body>
        </html>
        """
        sendResponse(fd, statusCode: 200, contentType: "text/html", body: html)
    }

    private func sendResponse(_ fd: Int32, statusCode: Int, contentType: String, body: String) {
        let reason = statusCode == 200 ? "OK" : statusCode == 404 ? "Not Found" : "Unknown"
        let response = "HTTP/1.0 \(statusCode) \(reason)\r\n" +
            "Content-Type: \(contentType)\r\n" +
            "Content-Length: \(body.utf8.count)\r\n" +
            "Connection: close\r\n" +
            "Access-Control-Allow-Origin: *\r\n\r\n" +
            body
        writeString(fd, response)
    }

    private func writeString(_ fd: Int32, _ string: String) {
        let data = string.data(using: .utf8)!
        data.withUnsafeBytes { ptr in
            if let base = ptr.baseAddress {
                _ = write(fd, base, data.count)
            }
        }
    }
}

File: android-app/app/src/main/java/com/rifatcam/pro/StreamServer.kt
package com.rifatcam.pro

import android.util.Log
import org.json.JSONObject
import java.io.IOException
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

class StreamServer(
    private val port: Int = NetworkUtils.STREAM_PORT,
    private val deviceName: String = "RifatCam"
) {
    interface FrameProvider {
        fun getLatestJpeg(): ByteArray?
        fun getInfoJson(): String
    }

    private var serverSocket: ServerSocket? = null
    private var running = false
    private var frameProvider: FrameProvider? = null
    private val clients = mutableListOf<ClientHandler>()
    private val tag = "StreamServer"

    fun setFrameProvider(provider: FrameProvider) {
        this.frameProvider = provider
    }

    fun start() {
        if (running) return
        running = true
        Thread({ serverLoop() }, "stream-server").apply {
            isDaemon = true
            start()
        }
        Log.i(tag, "Stream server started on port $port")
    }

    fun stop() {
        running = false
        try {
            serverSocket?.close()
        } catch (e: IOException) {
            // ignore
        }
        synchronized(clients) {
            for (client in clients) {
                client.stop()
            }
            clients.clear()
        }
        Log.i(tag, "Stream server stopped")
    }

    fun getClientCount(): Int {
        synchronized(clients) {
            return clients.size
        }
    }

    private fun serverLoop() {
        try {
            serverSocket = ServerSocket()
            serverSocket?.reuseAddress = true
            serverSocket?.bind(InetSocketAddress(port))
            Log.i(tag, "Server listening on 0.0.0.0:$port")

            while (running) {
                try {
                    val clientSocket = serverSocket?.accept() ?: continue
                    clientSocket.soTimeout = 10000
                    val handler = ClientHandler(clientSocket)
                    synchronized(clients) {
                        clients.add(handler)
                    }
                    handler.start()
                    Log.d(tag, "Client connected: ${clientSocket.inetAddress.hostAddress}")
                } catch (e: IOException) {
                    if (running) {
                        Log.e(tag, "Accept error: ${e.message}")
                    }
                }
            }
        } catch (e: IOException) {
            if (running) {
                Log.e(tag, "Server error: ${e.message}")
            }
        }
    }

    inner class ClientHandler(private val socket: Socket) : Thread("stream-client") {
        private var clientRunning = true

        fun stopClient() {
            clientRunning = false
            try {
                socket.close()
            } catch (e: IOException) {
                // ignore
            }
        }

        override fun run() {
            try {
                val inputStream = socket.getInputStream()
                val outputStream = socket.getOutputStream()

                val request = readRequest(inputStream)

                if (request == null) {
                    sendError(outputStream, 400, "Bad Request")
                    return
                }

                when {
                    request.startsWith("GET /video") -> handleVideoStream(outputStream)
                    request.startsWith("GET /info") -> handleInfo(outputStream)
                    request.startsWith("POST /control") -> {
                        val body = readBody(inputStream, request)
                        handleControl(outputStream, body)
                    }
                    request.startsWith("GET /") -> handleIndex(outputStream)
                    else -> sendError(outputStream, 404, "Not Found")
                }
            } catch (e: IOException) {
                Log.d(tag, "Client disconnected: ${e.message}")
            } finally {
                try {
                    socket.close()
                } catch (e: IOException) {
                    // ignore
                }
                synchronized(clients) {
                    clients.remove(this)
                }
            }
        }

        private fun readRequest(inputStream: java.io.InputStream): String? {
            try {
                val buffer = ByteArray(4096)
                val bytesRead = inputStream.read(buffer)
                if (bytesRead <= 0) return null
                val request = String(buffer, 0, bytesRead, StandardCharsets.UTF_8)
                val headerEnd = request.indexOf("\r\n")
                return if (headerEnd >= 0) request.substring(0, headerEnd) else null
            } catch (e: Exception) {
                return null
            }
        }

        private fun readBody(inputStream: java.io.InputStream, request: String): String {
            val contentLength = request.lines().find { it.lowercase().startsWith("content-length:") }
                ?.split(":")?.get(1)?.trim()?.toIntOrNull() ?: return "{}"
            try {
                val body = ByteArray(contentLength)
                var totalRead = 0
                while (totalRead < contentLength) {
                    val read = inputStream.read(body, totalRead, contentLength - totalRead)
                    if (read == -1) break
                    totalRead += read
                }
                return String(body, 0, totalRead, StandardCharsets.UTF_8)
            } catch (e: Exception) {
                return "{}"
            }
        }

        private fun handleVideoStream(outputStream: OutputStream) {
            val boundary = "--boundary"
            val contentType = "multipart/x-mixed-replace; boundary=$boundary"

            sendHeaders(outputStream, 200, contentType)

            val boundaryBytes = "--$boundary\r\n".toByteArray()
            val headerTemplate = "Content-Type: image/jpeg\r\nContent-Length: %d\r\n\r\n"

            while (clientRunning && running) {
                val jpegData = frameProvider?.getLatestJpeg()
                if (jpegData != null) {
                    try {
                        outputStream.write(boundaryBytes)
                        outputStream.write(headerTemplate.format(jpegData.size).toByteArray())
                        outputStream.write(jpegData)
                        outputStream.write("\r\n".toByteArray())
                        outputStream.flush()
                    } catch (e: IOException) {
                        break
                    }
                } else {
                    try {
                        Thread.sleep(30)
                    } catch (e: InterruptedException) {
                        break
                    }
                }
            }
        }

        private fun handleInfo(outputStream: OutputStream) {
            val info = frameProvider?.getInfoJson() ?: """{"status":"active","name":"$deviceName","port":$port}"""
            val body = info.toByteArray()
            sendHeaders(outputStream, 200, "application/json", body.size)
            outputStream.write(body)
            outputStream.flush()
        }

        private fun handleControl(outputStream: OutputStream, body: String) {
            try {
                val json = JSONObject(body)
                val command = json.optString("command", "")
                Log.d(tag, "Control command: $command")
                val response = """{"status":"ok","command":"$command"}"""
                val responseBytes = response.toByteArray()
                sendHeaders(outputStream, 200, "application/json", responseBytes.size)
                outputStream.write(responseBytes)
            } catch (e: Exception) {
                sendError(outputStream, 400, "Invalid JSON")
            }
        }

        private fun handleIndex(outputStream: OutputStream) {
            val html = """
                <!DOCTYPE html>
                <html>
                <head>
                    <title>RifatCam Pro - $deviceName</title>
                    <meta name="viewport" content="width=device-width, initial-scale=1">
                    <style>
                        * { margin: 0; padding: 0; box-sizing: border-box; }
                        body { background: #0a0e17; color: #fff; font-family: 'Segoe UI', sans-serif; display: flex; flex-direction: column; align-items: center; justify-content: center; min-height: 100vh; }
                        h1 { color: #00d4ff; font-size: 24px; margin-bottom: 20px; letter-spacing: 3px; }
                        img { width: 100%; max-width: 720px; border: 1px solid rgba(0,212,255,0.3); border-radius: 12px; }
                        .info { margin-top: 20px; color: #8899aa; font-size: 14px; }
                        .status { color: #00ff88; margin-top: 10px; }
                    </style>
                </head>
                <body>
                    <h1>◆ RIFATCAM PRO</h1>
                    <img src="/video" alt="Live Stream" />
                    <div class="info">Device: $deviceName | Port: $port</div>
                    <div class="status">● LIVE</div>
                </body>
                </html>
            """.trimIndent()
            val body = html.toByteArray()
            sendHeaders(outputStream, 200, "text/html", body.size)
            outputStream.write(body)
            outputStream.flush()
        }

        private fun sendHeaders(outputStream: OutputStream, statusCode: Int, contentType: String, contentLength: Int = -1) {
            val reason = when (statusCode) {
                200 -> "OK"
                400 -> "Bad Request"
                404 -> "Not Found"
                500 -> "Internal Server Error"
                else -> "Unknown"
            }
            val headers = StringBuilder()
            headers.append("HTTP/1.0 $statusCode $reason\r\n")
            headers.append("Content-Type: $contentType\r\n")
            headers.append("Connection: close\r\n")
            headers.append("Cache-Control: no-cache, no-store, must-revalidate\r\n")
            headers.append("Pragma: no-cache\r\n")
            headers.append("Access-Control-Allow-Origin: *\r\n")
            if (contentLength >= 0) {
                headers.append("Content-Length: $contentLength\r\n")
            }
            headers.append("\r\n")
            outputStream.write(headers.toString().toByteArray())
            outputStream.flush()
        }

        private fun sendError(outputStream: OutputStream, statusCode: Int, message: String) {
            val body = """{"error":"$message"}"""
            sendHeaders(outputStream, statusCode, "application/json", body.length)
            outputStream.write(body.toByteArray())
            outputStream.flush()
        }
    }
}
